copy/copy.go

```diff
@@ -548,7 +563,7 @@ func layerDigestsDiffer(a, b []types.BlobInfo) bool {
 
 // copyUpdatedConfigAndManifest updates the image per ic.manifestUpdates, if necessary,
 // stores the resulting config and manifest to the destination, and returns the stored manifest.
-func (ic *imageCopier) copyUpdatedConfigAndManifest(ctx context.Context) ([]byte, error) {
+func (ic *imageCopier) copyUpdatedConfigAndManifest(ctx context.Context, expectedDigest *digest.Digest) ([]byte, error) {
```

Comment:
Check for `expectedDigest` all the way down here is unnecessarily late.

Instead, please do it just after `image.FromUnparsedImage`, in the same block that sets `destisDigestedReference`.  That will mean that users submitting digests that don’t match the source will see an immediate failure instead of having to wait for all layers to be copied.

And if `!c.canModifyManifest`, `copyUpdatedConfigAndManifest` already uses `ic.src.Manifest`, which is exactly the source manifest unmodified, so an extra check that it matches the digest at this point is not necessary (if the manifest was checked before copying layers).

Reply:
Moved it.